Review of api/validators

mmalerba opened the pull request:
See individual commits which address the TODOs I initially commented

mmalerba committed:
Review of api/validators

mmalerba commented on packages/forms/experimental/src/api/validators/max.ts:
```diff
@@ -28,6 +33,7 @@ export function max<TPathKind extends PathKind = PathKind.Root>(
 
   metadata(path, MAX, reactiveMaxValue);
   validate(path, (ctx) => {
+    // TODO: resolve TODO below
```
I think probably just `NaN`? The rest should be prohibited by the type. For `NaN` I would say we should just treat it the same as `undefined` (i.e. it means there is no max)

mmalerba commented on packages/forms/experimental/src/api/validators/max_length.ts:
```diff
@@ -14,16 +14,24 @@ import {BaseValidatorConfig, ValueWithLength} from './types';
+ *    to be used instead of the default `ValidationError.maxlength(maxLength)`
+ * @template TPathKind The kind of path the logic is bound to (a root path, child path, or item of an array)
  */
+// TODO: We should decide whether we want to use camel case or all-lowercase, to create an error we
```
Do we prefer `minLength` or `minlength`?

mmalerba commented on packages/forms/experimental/src/api/validators/max_length.ts:
```diff
@@ -14,16 +14,24 @@ import {BaseValidatorConfig, ValueWithLength} from './types';
   maxLength: number | LogicFn<ValueWithLength, number | undefined, TPathKind>,
   config?: BaseValidatorConfig<ValueWithLength, TPathKind>,
 ) {
+  // TODO: should we memoize this in a computed? (applies to other metadata validators as well)
```
We don't know what the user is doing in their logic function, it could potentially be expensive so my feeling is yes we probably should memoize this?

mmalerba commented on packages/forms/experimental/src/api/validators/min_length.ts:
```diff
@@ -28,6 +33,7 @@ export function minLength<TPathKind extends PathKind = PathKind.Root>(
   metadata(path, MIN_LENGTH, reactiveMinLengthValue);
 
   validate(path, (ctx) => {
+    // TODO: resolve TODO below
```
I vote no, sets don't have a length, they have a size. I could see just supporting `{length: number}` though, as is suggested in another TODO somewhere

michael-small replied:
Relatively recently in the current forms API in [19.2 ](https://github.com/angular/angular/releases/tag/19.2.0), sets were accounted for with a helper `lengthOrSize` https://github.com/angular/angular/pull/45793/files#diff-9b7158180baa4cab9fa2e451531c658358a253e496809f3e305e1c15866e711cR38

mmalerba replied:
Ah thanks for pointing that out, I wasn't aware of that. Given that history I'm fine supporting size as well

mmalerba commented on packages/forms/experimental/src/api/validators/types.ts:
```diff
@@ -3,14 +3,17 @@
 import {LogicFn, PathKind, ValidationResult} from '../types';
 
+/** Represents a value that has a length. */
+// TODO: resolve TODO below
```
Yeah this feels reasonable to me

mmalerba committed:
improve ergonomics of errors option on validators

Improves the ergonomics of the `errors` option offered by the standard
validators.
1. Renames it to `error` because in the majority of cases it will be
   used to return a single `ValidationError`.
2. Allows directly using a constant `ValidationError` insetad of having
   to create a function like `() => ValidationError.custom()`

mmalerba committed:
handle NaN case in min/max

mmalerba committed:
switch ValidationError to use camelCase names

mmalerba committed:
memoize result of logic functions

mmalerba committed:
allow any value with length or size for minlength/maxlength

michael-small commented on packages/forms/experimental/src/api/validators/max.ts:
```diff
@@ -6,39 +6,48 @@
 
 /**
- * Validator requiring a field value to be smaller than or equal to a maximum value.
+ * Binds a validator to the given path that requires the the value to be less than or equal to the
```
```suggestion
 * Binds a validator to the given path that requires the value to be less than or equal to the
```

michael-small commented on packages/forms/experimental/src/api/validators/min.ts:
```diff
@@ -6,38 +6,48 @@
 
 /**
- * Validator requiring a field value to be greater than or equal to a minimum value.
+ * Binds a validator to the given path that requires the the value to be greater than or equal to
```
```suggestion
 * Binds a validator to the given path that requires the value to be greater than or equal to
```

mmalerba committed:
fix typos

Co-authored-by: michael-small <33669563+michael-small@users.noreply.github.com>

mmalerba commented:
/gemini review

mmalerba commented on packages/forms/experimental/src/api/validators/max.ts:
```diff
@@ -6,39 +6,48 @@
 ) {
-  const reactiveMaxValue = typeof maxValue === 'number' ? () => maxValue : maxValue;
+  const reactiveMaxValue = define(path, (ctx) => {
+    return computed(() => (typeof maxValue === 'number' ? maxValue : maxValue(ctx)));
```
We actually wouldn't want to do that. The reactive keys aggregates its values (so in this case takes the minimum max of all the max rules defined for this field across the schema).

I agree that this is super confusing and its one of the reasons I don't like the current naming of these things (see my naming proposal in slack)

mmalerba replied:
And yes, I am aware of the fact that I'm the one who named it in the first place 🤣

mmalerba committed:
address feedback
